### DISCUSSION TITLE
Loading SC ATS/RTS tables?

### DISCUSSION THREAD

**Discussion by Dr-DW** [nasa/cFS#499] [Q&A]
I am using OpenSatKit's Mission FSW only with cFS for now. I am trying to view, edit, load, and activate stored commands. I want to learn and use both ATS and RTS. I found good stored command documentation SC/docs/user_guide_historical. It states the following:

"To prepare an ATS load, follow these steps:
1.	Plan the sequence to be loaded. Address the spacecraft configuration, commands, times for the commands, size of the commands, and other SC activity.
2.	Map out the ATS in Systems Test and Operations Language (STOL) Command Mnemonics with the appropriate configured time tags.
3.	Translate the STOL command mnemonic procedure into a loadable file by the FOT."

FOT stands for "Flight Operations team", which I guess is me. Anyways, all of this to say I found some example files "sc_ats1.tbl" and "sc_rts001.tbl" (1-8 files) in the OpenSatKit-master/cfs/build/exe/simsat/cf directory. When trying to display the tables in Table Manager it takes me to OpenSatKit-master/cosmos/config/tools/table_manager looking for a table definition text file. None of them match the SC table files. I tried loading another random file "cs_eeprom.tbl" which had a matching definition and it worked.

I tried looking elsewhere online and couldn't find anything. Also at the end of the documentation above it states the following:
"Documentation of ATS Definition Tables is located at (mission defined)." and
"Documentation of RTS Definition Tables is located at (mission defined)."

TLDR; Where might I find some SC ATS/RTS Definitions? Or how can I read and edit the tables? Thanks!

**Comment by dmccomas** [comment]
Unfortunately I don't have solutions for everything, but hopefully this reply will clarify the situation.

As you confirmed with the checksum table, Table Manager uses definition files to understand how to interpret the contents of a binary file. The definition files in OSK are for files/tables with a fixed binary format. Some of these 'fixed formats' can have variable length arrays of fixed records definitions so you'll get warnings if the array length doesn't equal the max length defined in Table Manager definition file.    

ATS and RTS table files contain sequences of commands stored as binary so they have a variable format.  A more sophisticated tool than Table Manager is required that can interpret each command in the file based on it's command message identifier. COSMOS has a tool called command sequencer that generates text-based command sequences. This tool could be augmented to generate binary files. I'm not aware of anyone releasing a COSMOS ATS/RTS generation/display tool. 

All tables can be defined using C source files and the cmake build system, although this may become impractical based on the size and complexity of your ATS and RTS tables. The default OSK ATS and RTS tables you mentioned get loaded during initialization and can be executed.. 

The SC documentation states 'mission defined' because a mission's ground system's command & telemetry definitions are required as input to a tool that can generate and display ATS/RTS tables.

**Reply by Dr-DW** [reply]
Thank you! That was super helpful. I have tried the command sequencer and am happy with how it works. Can you point me in the direction for turning the text file into a binary file? I don't know if there is a special format it needs before converting.

Edit: Looking further into the documentation the ATS and RTS have different Binary formats. I am not very familiar with the structure of cFS / OpenSatKit. Where would the source / build files be for ATS/RTS tables?

Edit 2 for anyone else interested: Found this documentation on making tables [link](https://github.com/nasa/elf2cfetbl/blob/main/Guide-Creating-Table-files.txt) and OpenSatKit-master/cfs/apps/sc/fsw/src/sc_tbldefs.h holds the definitions. OpenSatKit-master/cfs/apps/sc/fsw/src/tables holds the C files needed for compilation.
